### PR TITLE
gdbstub:we can use uart to debug with gdb when crash

### PR DESCRIPTION
**Pull request by anjiahao1**
## Summary
we can use uart to debug with gdb when crash
## Impact
gdbstub impove speed by
1. expand buffer size
2. RLE compression
## Testing

mps3 an547 with gdbstub

**MPS3 AN547 with GDB Stub Testing Description**
`SERIAL_GDBSTUB_AUTO_ATTACH`
This option automatically attaches the serial port to the GDB stub. When enabled, the serial port specified by GDBSTUB_SERIAL_PATH is dedicated exclusively to the GDB stub. This ensures that the GDB debugger can be connected seamlessly without needing manual intervention to configure the serial port.

`SERIAL_GDBSTUB_PANIC_TIMEOUT`
This setting introduces a delay before the GDB stub connection times out during a system panic. The value defines the timeout duration, allowing the debugger enough time to attach when a crash occurs. A value of 0 disables the timeout, meaning the system will wait indefinitely for a connection during a panic.

`BOARD_MEMORY_RANGE`
This option defines the memory range accessible by GDB. By configuring BOARD_MEMORY_RANGE, you can control which parts of the system memory GDB is permitted to access, ensuring both secure and efficient debugging.

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



This PR description needs significant improvement to meet NuttX requirements. Here's a breakdown:

**Missing/Unclear Information:**

* **What specific problem does this solve?**  Simply stating "we can use uart to debug" isn't helpful. Is the current GDB stub broken? Is this a new feature entirely? 
* **What code is changed?**  Mentioning "gdbstub" is too vague. Which files are modified? 
* **How does the buffer size improvement work?** What was the old size, what is the new size, and why is this faster?
* **No issue references:** Are there related NuttX issues this PR addresses?
* **Incomplete testing:** 
    * "mps3 an547" is insufficient. Provide the full NuttX configuration used.
    * What are the testing logs? Just stating "with gdbstub" doesn't show what was tested or if the change worked.

**How to Improve:**

1. **Specific Problem:** Clearly state the problem this PR solves. For example:
    * "The current GDB stub fails to connect reliably on UART" 
    * "NuttX lacks a way to debug kernel panics over UART"

2. **Detailed Changes:**  Be specific about code modifications:
    * "Modified `drivers/serial/gdbstub.c` to increase the input buffer size from X to Y bytes"
    * "Added new configuration option `SERIAL_GDBSTUB_AUTO_ATTACH` to..."

3. **Explain the "Why":** 
    * "Increasing the buffer size improves GDB communication speed by reducing the number of UART transactions."

4. **Complete Testing Information:**
    * List ALL build hosts and targets used for testing (OS, architecture, compiler versions).
    * Include REAL testing logs, showing the issue before and the successful behavior after the change.

**Example Summary:**

> This PR enhances the NuttX GDB stub to improve debugging reliability and speed over UART connections. It introduces a new configuration option, `SERIAL_GDBSTUB_AUTO_ATTACH`, to automatically attach the stub to a specified serial port upon startup. Additionally, the input buffer size within `gdbstub.c` has been increased from 64 to 256 bytes, significantly reducing UART transaction overhead and improving communication speed during debugging sessions.

Remember, a good PR description is clear, concise, and provides all the information reviewers need to understand and evaluate the change.

File: libs/libc/gdbstub/lib_gdbstub.c [hunk 1]
```diff
@@ -1786,6 +1939,16 @@ FAR struct gdb_state_s *gdb_state_init(gdb_send_func_t send,
   state->priv = priv;
   state->monitor = monitor;
 
+  if (CONFIG_BOARD_MEMORY_RANGE[0] != '\0')
+    {
+      state->range = alloc_memory_region(CONFIG_BOARD_MEMORY_RANGE);
```
Comment (anchao) [inline libs/libc/gdbstub/lib_gdbstub.c:1944]: could we remove all parser/allocate code and declare `CONFIG_BOARD_MEMORY_RANGE` as array format? 
`static const struct memory_region_s g_memory_regions[] = CONFIG_BOARD_MEMORY_RANGE;`
which will be beneficial for identifying bring up issues.

**Review by anchao — CHANGES_REQUESTED** [review]

File: libs/libc/gdbstub/lib_gdbstub.c (hunk 1 above)
Comment (anjiahao1) [reply libs/libc/gdbstub/lib_gdbstub.c:1944]: Kconfig how to define a struct array？

Comment (anchao) [reply libs/libc/gdbstub/lib_gdbstub.c:1944]: > Kconfig how to define a struct array？

@anjiahao1 

image

image

FYI @xiaoxiang781216 @XuNeo

Comment (xiaoxiang781216) [reply libs/libc/gdbstub/lib_gdbstub.c:1944]: parse code is already added before this patch, so it's better to create a new patch instead.

Comment (anchao) [reply libs/libc/gdbstub/lib_gdbstub.c:1944]: All static resources could be defined in this style. Using dynamic allocator will make these debugging features unusable in some extreme scenarios.

Comment (xiaoxiang781216) [reply libs/libc/gdbstub/lib_gdbstub.c:1944]: Yes, but it should belong to a new patch to remove ALL places.

Comment (anjiahao1) [reply libs/libc/gdbstub/lib_gdbstub.c:1944]: Great! @anchao

File: drivers/misc/dev_mem.c [hunk 2]
```diff
@@ -34,6 +34,11 @@
 
 #define DEVMEM_REGION 8
 
+#ifdef CONFIG_BOARD_MEMORY_RANGE
+static const struct memory_region_s g_memory_region[] =
+  CONFIG_BOARD_MEMORY_RANGE;
```
Comment (xiaoxiang781216) [inline drivers/misc/dev_mem.c:39]: let's add {} and remove it from line 4899

File: sched/misc/coredump.c [hunk 3]
```diff
@@ -51,6 +51,10 @@ static struct lib_blkoutstream_s  g_blockstream;
 static unsigned char *g_blockinfo;
 #endif
 
+#ifdef CONFIG_BOARD_MEMORY_RANGE
+static struct memory_region_s g_memory_region[] =
```
Comment (xiaoxiang781216) [inline sched/misc/coredump.c:55]: add const

File: libs/libc/gdbstub/lib_gdbstub.c [hunk 4]
```diff
@@ -48,27 +50,33 @@
...
+                                            */
+  int last_stopreason;                     /* Last stop reason */
+  FAR void *last_stopaddr;                 /* Last stop address */
+  pid_t pid;                               /* Gdb current thread */
+  FAR char *pkt_next;                      /* Pointer to next byte in packet */
+  char pkt_buf[1024];                      /* Packet buffer */
+  size_t pkt_len;                          /* Packet send and receive length */
+  uint8_t running_regs[XCPTCONTEXT_SIZE];  /* Registers of running thread */
+  size_t size;                             /* Size of registers */
+  FAR const struct memory_region_s *range; /* Memory regions */
```
Comment (xiaoxiang781216) [inline libs/libc/gdbstub/lib_gdbstub.c:80]: why not use g_memory_region directly

**Comment by xiaoxiang781216** [comment]
let's move the dropping parse code to the last patch.

**Comment by JianyuWang0623** [comment]
@anjiahao1 @xiaoxiang781216  Could you sync 23d70a87551d26fb0a8a3e9e26443e8db412331a to internal gerrit pleasee? 
@GUIDINGLI is checking progress of upstream.

**Pull request merged**